### PR TITLE
[IMP] account: allow to reconcile taking amount from the label of the statement line

### PR DESCRIPTION
**Pull request by william-andre**
Task 2039772

Add a new amount matching mode by taking the amount from the label, based on a regex. The regex must contain a capture group. If multiple groups are matched, there will be an error.
Also add a 'remaining' option for the second line, that puts the remaining amount of the statement in the second line.

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_reconcile_model.py [hunk 1]
```diff
@@ -92,12 +94,15 @@ class AccountReconcileModel(models.Model):
     second_label = fields.Char(string='Second Journal Item Label')
     second_amount_type = fields.Selection([
         ('fixed', 'Fixed'),
-        ('percentage', 'Percentage of amount')
+        ('percentage', 'Percentage of amount'),
```
Comment (csnauwaert) [inline addons/account/models/account_reconcile_model.py]: rename: Percentage of balance and remove the remaining options. In order to use the remaining user will have to create a second line with percentage 100% of balance.

File: addons/account/models/account_reconcile_model.py [hunk 2]
```diff
@@ -74,12 +74,14 @@ class AccountReconcileModel(models.Model):
     label = fields.Char(string='Journal Item Label')
     amount_type = fields.Selection([
         ('fixed', 'Fixed'),
-        ('percentage', 'Percentage of balance')
+        ('percentage', 'Percentage of balance'),
+        ('regex', 'From label'),
```
Comment (csnauwaert) [inline addons/account/models/account_reconcile_model.py]: You should also add the support for regex in def _get_write_off_move_lines_dict

File: addons/account/static/src/js/reconciliation/reconciliation_model.js [hunk 3]
```diff
@@ -1126,7 +1126,26 @@ var StatementModel = BasicModel.extend({
...
+                break;
+            case 'regex':
+                var matching = line.st_line.name.match(new RegExp(values.amount_from_label_regex))
+                amount = parseFloat(!matching || matching.length != 2 ? null : matching[1].replace(',', '.').replace(/\.(?=.*\.)/, ""));
+                break;
+            case 'fixed':
+                amount = values.amount;
+                break;
+            case 'remaining':
+                amount = line.balance.amount - line.reconciliation_proposition.slice(-1)[0].amount;
```
Comment (csnauwaert) [inline addons/account/static/src/js/reconciliation/reconciliation_model.js]: In quickCreateProposition: function (handle, reconcileModelId)
Change the way it works to match the way it works in python with auto reconcile -> Create first line, then call this._computeLine(line) then create the second line if there is any and finally call _computeLine(line) again.

And you can remove the remaining case

File: addons/account/static/src/js/reconciliation/reconciliation_model.js [hunk 4]
```diff
@@ -1126,7 +1126,26 @@ var StatementModel = BasicModel.extend({
...
         };
-        var amount = values.amount !== undefined ? values.amount : line.balance.amount;
+        var amount;
+        switch(values.amount_type) {
+            case 'percentage':
+                amount = line.balance.amount * values.amount / 100;
+                break;
+            case 'regex':
+                var matching = line.st_line.name.match(new RegExp(values.amount_from_label_regex))
+                amount = parseFloat(!matching || matching.length != 2 ? null : matching[1].replace(',', '.').replace(/\.(?=.*\.)/, ""));
```
Comment (csnauwaert) [inline addons/account/static/src/js/reconciliation/reconciliation_model.js]: I'm not entirely a fan of this as what would happen if the thousand separator is . and we have something like this 100.000 (without any decimals) but I guess that this is a corner case scenario so I can live with it

Comment (william-andre) [reply addons/account/static/src/js/reconciliation/reconciliation_model.js]: added the decimal separator as a parameter

File: addons/account/static/tests/reconciliation_tests.js [hunk 5]
```diff
@@ -192,8 +192,8 @@ var db = {
...
+            {'second_analytic_account_id': false, 'second_amount_type': "percentage", 'second_journal_id': false, 'id': 3, 'analytic_account_id': false, 'display_name': "ATOS", 'rule_type': 'writeoff_button', 'second_tax_ids': [7], 'has_second_line': true, 'journal_id': false, 'label': "ATOS Banque", 'second_label': "ATOS Frais", 'second_account_id': 286, 'account_id': 285, 'company_id': [1, "Demo SPRL"], 'tax_ids': [6], 'amount_type': "percentage", 'name': "ATOS", 'amount': 97.5, 'second_amount': -14.75},
+            {'second_analytic_account_id': false, 'second_amount_type': "percentage", 'second_journal_id': false, 'id': 10, 'analytic_account_id': false, 'display_name': "Double", 'rule_type': 'writeoff_button', 'second_tax_ids': [], 'has_second_line': true, 'journal_id': false, 'label': "Double Banque", 'second_label': "Double Frais", 'second_account_id': 286, 'account_id': 285, 'company_id': [1, "Demo SPRL"], 'tax_ids': [], 'amount_type': "percentage", 'name': "Double", 'amount': 97.5, 'second_amount': 100, 'match_journal_ids': [], 'analytic_tag_ids': [1,2]},
```
Comment (qdp-odoo) [inline addons/account/static/tests/reconciliation_tests.js]: the whole diff of this file is not really understandable

File: addons/account/models/chart_template.py [hunk 6]
```diff
@@ -1154,9 +1154,12 @@ class AccountReconcileModelTemplate(models.Model):
     label = fields.Char(string='Journal Item Label')
     amount_type = fields.Selection([
         ('fixed', 'Fixed'),
-        ('percentage', 'Percentage of balance')
+        ('percentage', 'Percentage of balance'),
+        ('regex', 'From label'),
         ], required=True, default='percentage')
     amount = fields.Float(string='Write-off Amount', digits=0, required=True, default=100.0, help="Fixed amount will count as a debit if it is negative, as a credit if it is positive.")
+    amount_from_label_regex = fields.Char()
```
Comment (qdp-odoo) [inline addons/account/models/chart_template.py]: default = (\d+[,\.]?\d+)

File: addons/account/models/chart_template.py [hunk 7]
```diff
@@ -1154,9 +1154,12 @@ class AccountReconcileModelTemplate(models.Model):
...
+    decimal_separator = fields.Char(help="Every character that is nor a digit nor this separator will be removed from the matching string")
```
Comment (qdp-odoo) [inline addons/account/models/chart_template.py]: can't you just ignore `,` and `.` all the time and let parseFloat recognize thousand separators and decimal points?

File: addons/account/models/account_reconcile_model.py [hunk 8]
```diff
@@ -92,12 +94,15 @@ class AccountReconcileModel(models.Model):
...
         ], required=True, default='percentage')
     show_force_tax_included = fields.Boolean(store=False, help='Technical field used to show the force tax included button')
     force_tax_included = fields.Boolean(string='Tax Included in Price',
         help='Force the tax to be managed as a price included tax.')
     amount = fields.Float(string='Write-off Amount', digits=0, required=True, default=100.0, help="Fixed amount will count as a debit if it is negative, as a credit if it is positive.")
+    amount_from_label_regex = fields.Char(help="There is no need for regex delimiter, only the regex is needed. For instance if you want to extract the amount from\nR:9672938 10/07 AX 9415126318 T:5L:NA BRT: 3358,07 C:\nYou could enter\nBRT: ([\d,]+)")
+    decimal_separator = fields.Char(default=lambda self: self.env['res.lang']._lang_get(self.env.user.lang).decimal_point, help="Every character that is nor a digit nor this separator will be removed from the matching string")
```
Comment (qdp-odoo) [inline addons/account/models/account_reconcile_model.py]: see my remark on the template object

File: addons/account/models/account_reconcile_model.py [hunk 9]
```diff
@@ -92,12 +94,15 @@ class AccountReconcileModel(models.Model):
...
         ('fixed', 'Fixed'),
-        ('percentage', 'Percentage of balance')
+        ('percentage', 'Percentage of balance'),
+        ('regex', 'From label'),
         ], required=True, default='percentage')
     show_force_tax_included = fields.Boolean(store=False, help='Technical field used to show the force tax included button')
     force_tax_included = fields.Boolean(string='Tax Included in Price',
         help='Force the tax to be managed as a price included tax.')
     amount = fields.Float(string='Write-off Amount', digits=0, required=True, default=100.0, help="Fixed amount will count as a debit if it is negative, as a credit if it is positive.")
+    amount_from_label_regex = fields.Char(help="There is no need for regex delimiter, only the regex is needed. For instance if you want to extract the amount from\nR:9672938 10/07 AX 9415126318 T:5L:NA BRT: 3358,07 C:\nYou could enter\nBRT: ([\d,]+)")
```
Comment (qdp-odoo) [inline addons/account/models/account_reconcile_model.py]: see my remark on the template object

File: addons/account/models/account_reconcile_model.py [hunk 10]
```diff
@@ -110,12 +115,14 @@ class AccountReconcileModel(models.Model):
...
         ('fixed', 'Fixed'),
-        ('percentage', 'Percentage of amount')
+        ('percentage', 'Percentage of balance'),
+        ('regex', 'From label'),
         ], string="Second Amount type",required=True, default='percentage')
     show_second_force_tax_included = fields.Boolean(store=False, help='Technical field used to show the force tax included button')
     force_second_tax_included = fields.Boolean(string='Second Tax Included in Price',
         help='Force the second tax to be managed as a price included tax.')
     second_amount = fields.Float(string='Second Write-off Amount', digits=0, required=True, default=100.0, help="Fixed amount will count as a debit if it is negative, as a credit if it is positive.")
+    second_amount_from_label_regex = fields.Char()
```
Comment (qdp-odoo) [inline addons/account/models/account_reconcile_model.py]: add label and default value

File: addons/l10n_generic_coa/demo/account_reconcile_model.xml [hunk 11]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data noupdate="0">
+        <record id="reconcile_from_label" model="account.reconcile.model">
+            <field name="name">Line with Bank Fees</field>
+            <field name="rule_type">writeoff_suggestion</field>
+            <field name="match_label">contains</field>
+            <field name="match_label_param">BRT</field>
+            <field name="label">Due amount</field>
+            <field name="account_id" search="[('internal_group', '=', 'receivable'), ('internal_type', '=', 'asset')]"/>
```
Comment (qdp-odoo) [inline addons/l10n_generic_coa/demo/account_reconcile_model.xml]: account_id was empty in my test DB.

File: addons/l10n_generic_coa/demo/account_reconcile_model.xml [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+            <field name="amount_type">regex</field>
+            <field name="amount_from_label_regex">BRT: ([\d,]+)</field>
+            <field name="decimal_separator">,</field>
+            <field name="has_second_line" eval="True"/>
+            <field name="second_label">Bank Fees</field>
+            <field name="second_account_id" search="[('internal_group', '=', 'other'), ('internal_type', '=', 'expense')]"/>
```
Comment (qdp-odoo) [inline addons/l10n_generic_coa/demo/account_reconcile_model.xml]: second_account_id was empty in my test DB

File: addons/l10n_generic_coa/demo/account_reconcile_model.xml (hunk 11 above)
Comment (william-andre) [reply addons/l10n_generic_coa/demo/account_reconcile_model.xml]: It works now

**Comment by qdp-odoo** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**